flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendLoadingTest.java

```diff
@@ -0,0 +1,274 @@
+public class ChangelogStateBackendLoadingTest {
+    @Rule public final TemporaryFolder tmp = new TemporaryFolder();
+
+    @Parameterized.Parameters
+    public static List<Boolean> loadFromConfigOnly() {
+        return Arrays.asList(true, false);
+    }
+
+    @Parameterized.Parameter public boolean configOnly;
+
+    private final ClassLoader cl = getClass().getClassLoader();
+
+    private final String backendKey = CheckpointingOptions.STATE_BACKEND.key();
+
+    @Test
+    public void testLoadingDefault() throws Exception {
+        final StateBackend backend =
+                StateBackendLoader.loadStateBackendFromApplicationOrConfigOrDefault(
+                        null, config(), cl, null);
+        final CheckpointStorage storage =
+                CheckpointStorageLoader.load(null, null, backend, config(), cl, null);
+
+        assertDelegateStateBackend(
+                backend, HashMapStateBackend.class, storage, JobManagerCheckpointStorage.class);
+    }
+
+    @Test
+    public void testLoadingMemoryStateBackendFromConfig() throws Exception {
+        final Configuration config = config("jobmanager");
+        final StateBackend backend = loadStateBackend(config);
+        final CheckpointStorage storage =
+                CheckpointStorageLoader.load(null, null, backend, config, cl, null);
+
+        assertDelegateStateBackend(
+                backend, MemoryStateBackend.class, storage, MemoryStateBackend.class);
+    }
+
+    @Test
+    public void testLoadingFsStateBackendFromConfig() throws Exception {
+        final Configuration config = config("filesystem");
+        final StateBackend backend = loadStateBackend(config);
+        final CheckpointStorage storage =
+                CheckpointStorageLoader.load(null, null, backend, config, cl, null);
+
+        assertDelegateStateBackend(
+                backend, HashMapStateBackend.class, storage, JobManagerCheckpointStorage.class);
+    }
+
+    @Test
+    public void testLoadingHashMapStateBackendFromConfig() throws Exception {
+        final Configuration config = config("hashmap");
+        final StateBackend backend = loadStateBackend(config);
+        final CheckpointStorage storage =
+                CheckpointStorageLoader.load(null, null, backend, config(), cl, null);
+
+        assertDelegateStateBackend(
+                backend, HashMapStateBackend.class, storage, JobManagerCheckpointStorage.class);
+    }
+
+    @Test
+    public void testLoadingRocksDBStateBackendFromConfig() throws Exception {
+        final Configuration config = config("rocksdb");
+        final StateBackend backend = loadStateBackend(config);
+        final CheckpointStorage storage =
+                CheckpointStorageLoader.load(null, null, backend, config(), cl, null);
+
+        assertDelegateStateBackend(
+                backend,
+                EmbeddedRocksDBStateBackend.class,
+                storage,
+                JobManagerCheckpointStorage.class);
+    }
```

Comment:
This code is duplicated many times in this class.
I think duplication can be eliminated by extracting test parameters (backend name, backend class, storage class).
But as I said above, I don't think we need to test all these combinations.

Reply:
The main reason I've added all types of backend loading tests is because of the separation of checkpoint storage and statebackend (and all the default settings).

I want to make sure all types are correct. If later we deprecated and simplify rocksdb/fs/mem statebackend, I think these tests can be removed.

But I will try to deduplicate this.